File: 007/main.js
/**
 * @param {number} x
 * @return {number}
 */
var reverse = function (x) {
  if (x === 0) return x
  let str = x + ''
  let flag = false
  let rev = []
  let doc
  if (x < 0) {
    flag = true
    let mid = str.split('')
    mid.splice(0, 1)
    rev = mid.reverse()
  } else {
    flag = false
    rev = str.split('').reverse()
  }
  while (rev[0] === '0') {
    rev.splice(0, 1)
  }
  let res = rev.join('')
  if (flag) {
    doc = parseInt('-' + res)
    if (doc + Math.pow(2, 31) < 0) {
      return 0
    }
    return doc
  } else {
    doc = parseInt(res)
    if (doc > Math.pow(2, 31) - 1) {
      return 0
    } else {
      return doc
    }
  }
}

console.log(reverse(123))
console.log(reverse(-123))
console.log(reverse(120))
console.log(reverse(0))

File: 007/main.cpp
#include <iostream>
#include <vector>
#include <math.h>
using namespace std;

class Solution {
public:
    int reverse(int x) {
        long long P1 = pow(2, 31) - 1;
        long long P2 = -pow(2, 31);
        if (x == 0) return 0;
        bool flag = false;
        if (x < 0) {
            x = -x;
            flag = true;
        }
        vector<int> v;
        int a;
        while (x) {
            a = x % 10;
            v.push_back(a);
            x = x / 10;
        }
        long long ans = 0;
        for (int i = 0; i < v.size(); i++) {
            if (i == 0 && v[i] == 0) continue;
            ans = ans * 10 + v[i];
        }
        if (flag) ans = -ans;
        if (ans > P1 || ans < P2) return 0;
        return ans;
    }
};

int main()
{
    Solution sol;
    int ans = sol.reverse(1147483647);
    cout << ans << endl;
    return 0;
}
